### ISSUE TITLE
qudt:isScalingOf not applied consistently

### ISSUE THREAD

**Issue by fkleedorfer** [qudt/qudt-public-repo#455]
I had high hopes to find the base unit for any scaled unit, but it seems it's not in the data.

E.g.
```
unit:DeciM qudt:isScalingOf unit:M ;
unit:CentiM qudt:isScalingOf unitM;
```
but
the triple is missing for 
```
unit:KiloM 
unit:MilliM 
```
Many other examples can be found. 

Having complete scaling and prefix information would be very useful, it also seems that the triples can be generated automatically based on the unit's localname.

**Comment by fkleedorfer** [comment]
Here's a CONSTRUCT query that generates all the missing scaling/prefix triples (I believe), for base units.
Note:
* Information measures (`unit:Byte`) etc are not included because the base units are selected by their multiplier of 1, but byte has a different one.
* Weights are arguably misleading as the base SI unit is `unit:KiloGM`, so all prefixes of scaled units have a wrong multiplier as using them would convert toward grams, not kilograms

Anyway: 
```
PREFIX qudt: <http://qudt.org/schema/qudt/>
PREFIX unit: <http://qudt.org/vocab/unit/>
PREFIX prefix: <http://qudt.org/vocab/prefix/>
PREFIX kind: <http://qudt.org/vocab/quantitykind/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
    ?scaledUnit
        qudt:isScalingOf ?baseUnit ;
        qudt:prefix ?prefix .
}
WHERE
{
    {
        # select all base units (multiplier 1, not using prefixes, except for KiloGM)
        ?baseUnit a qudt:Unit ;
            qudt:conversionMultiplier 1.0 .
            BIND(REPLACE(STR(?baseUnit), "^.+/", "") as ?baseUnitLocalName)
            FILTER(! CONTAINS(?baseUnitLocalName, "-"))
    }

    ?scaledUnit a qudt:Unit ;
    BIND(REPLACE(STR(?scaledUnit), "^.+/", "") as ?scaledUnitLocalName)
    ?prefix2
        a qudt:Prefix ;
        rdfs:label ?prefixLabel2 .
    FILTER EXISTS {

         FILTER (
            ?scaledUnitLocalName = CONCAT(STR(?prefixLabel2), ?baseUnitLocalName)
            || ?scaledUnitLocalName = REPLACE(?baseUnitLocalName, "KiloGM", CONCAT(STR(?prefixLabel2),"GM"))
            || ( ?scaledUnitLocalName = REPLACE(?baseUnitLocalName, "KiloGM", "GM") )
         )
    }
    FILTER(?scaledUnit != ?baseUnit)
    FILTER NOT EXISTS {
        ?scaledUnit qudt:isScalingOf ?unit
    }
    BIND ( IF( ?scaledUnit != unit:GM, ?prefix2, ?unbound) as ?prefix)
}
```

**Comment by steveraysteveray** [comment]
This is much appreciated! Look for the results to be included in the next release. Looks like it generates 100 new triples.

**Comment by fkleedorfer** [comment]
Ok well I've got another query in the works that connects all the scaled derived units (e.g. GM-PER-MilliM2) with their base units. Not quite there yet, but will have it tomorrow. Unless it's somehow a bad idea.

**Comment by steveraysteveray** [comment]
Your latest suggestion bears more thought. My question is whether there is a good use case for this new relation? (...and what would be a good name for the relation)?

**Comment by steveraysteveray** [comment]
Another big issue is that while we are **_defaulting_** to the SI system, the QUDT ontology also supports other systems of units of measure, which have different base units. So to be general, we would need to think about how to point back to these different base units in an intuitive way.

**Comment by fkleedorfer** [comment]
The use case is: What do I get when I divide a value of that unit by the conversionMultiplier? I know it's intended for converting between two units, but if I just convert to the base unit - which I can - what unit do I get? The `prefix` does not make so much sense there though

**Comment by fkleedorfer** [comment]
Related: I am a bit suprised that qudt does not allow me to find out the constituent units of derived units. E.g. for `unit:MilliGM-PER-M3` I'd expect links to `unit:MilliGM` and `unit:M3` - ideally with exponents. I know I can work those out from the dimension vectors, but I should not have to

**Comment by fkleedorfer** [comment]
.. or maybe in that case, I'd expect a link to  `unit:MilliGM` and `unit:M` - that way, the dimension vector in `unit:MilliGM-PER-M3` is directly applicable to the constituent units by row*column vector multiplication (otherwise, if `unit:M3` was linked, the exponent `L-3` could not be used directly)

**Comment by fkleedorfer** [comment]
... one way to do it would be along these lines:
```
unit:MilliGM-PER-M3
     qudt:component [
          qudt:constituentUnit unit:MilliGM ;
          qudt:exponent 1
     ] ;  
     qudt:component [
           qudt:constituentUnit unit:M;
           qudt:exponent -3
     ] ;
```
--
Another way to to it would be to just add 
```
unit:MilliGM-PER-M3
          qudt:constituentUnit unit:MilliGM ;
          qudt:constituentUnit unit:M;
```
.. and work out the exponents with some string magic on the dimension vectors whenever needed. Less data, more work in the queries.

**Comment by fkleedorfer** [comment]
Use case: I want to convert to/from other systems of units that represent their derived units as structs with base units and exponents. So I have to be able to find a qudt unit by base units and exponents, and I have to be able to know base units and exponents of a given qudt unit so as to generate the unit in that other system

**Comment by dr-shorthair** [comment]
These changes would move QUDT to support more directly some of functions that OM focuses on.

**Comment by fkleedorfer** [comment]
> These changes would move QUDT to support more directly some of functions that OM focuses on.

Is that good or bad?

**Comment by fkleedorfer** [comment]
By adding the `Y qudt:isScalingOf kiloGM` triples, we've established that `A qudt:isScalingOf B` means that B is what you get if you multiply a value of A with the `conversionMultiplier` and then subtract the `conversionOffset`. It does not mean, A is B with some prefix. 

So, we' ve got the case of `unit:L`, which is defined as 0.001 of a `unit:M3`, therefore, all scaled liter units, such as `unit:MilliL` etc. should get `qudt:isScalingOf unit:M3`, as should `unit:L`. Is that still compatible with what the `qudt:isScalingOf` property was introduced for? The property does not have a comment in the qudt schema, so it could go either way

**Comment by fkleedorfer** [comment]
After giving it some thought, I suggest the following:

Differentiate the two cases (and fix the changes from the last PR) such that they are consistent with previous versions of the ontology
1. `qudt:isScalingOf` just connects a unit with a prefix to the unit without the prefix. e.g.:
```
unit:KiloM qudt:isScalingOf unit:M .
unit:KiloGM qudt:isScalingOf unit:GM .
unit:MilliL qudt:isScalingOf unit:L .
```
and also
```
unit:DeciM3-PER-SEC qudt:isScalingOf unit:M3-PER-SEC
unit:MilliL-PER-SEC qudt:isScalingOf unit:L-PER-SEC
```
2. introduce e.g., `qudt:convertsTo` that links a unit with the unit you get when you apply the conversion formula. e.g:
```
unit:KiloM qudt:convertsTo unit:M .
unit:KiloGM qudt:convertsTo unit:KiloGM . #that one could be omitted bc it is the base unit, but it makes it easier if it's always there
unit:MilliL qudt:convertsTo unit:M3 .
```
and
```
unit:DeciM3-PER-SEC qudt:convertsTo unit:M3-PER-SEC
unit:MilliL-PER-SEC qudt:convertsTo unit:M3-PER-SEC
```

**Comment by fkleedorfer** [comment]
I found some chained uses of `qudt:isScalingOf` - is this a bug or is it intended`? 
``` 
 unit:KiloBIT-PER-SEC  qudt:isScalingOf unit:BIT-PER-SEC     
 unit:KiloBYTE-PER-SEC qudt:isScalingOf unit:BIT-PER-SEC     
 unit:MegaBIT-PER-SEC  qudt:isScalingOf unit:KiloBIT-PER-SEC 
```
```
 unit:PA-PER-HR       qudt:isScalingOf  unit:PA-PER-MIN      
 unit:PA-PER-MIN    qudt:isScalingOF unit:PA-PER-SEC   
```

**Comment by fkleedorfer** [comment]
So, my use case for conversion to another system was: convert something like

`{ { "unit" : "metre", "exponent", -3}, {"unit": "kilogram", "exponent", 1 }`  (never mind the json syntax I just invented.)

to 

`unit:KiloGM-PER-M3`

For that, there is still something missing (given all of the above mentioned triples are addes): I have to be able to determine that `unit:M3` is derived from `unit:M` with exponent 3. I suppose there are similar string-tricks I can do as with the prefixes, so this could be determined automatically (I'll try). I would connect the units using 

`qudt:baseUnit`

 and add 

`qudt:exponent`

, which must be the same as the exponent in the dimension vector. What do you think?

**Comment by fkleedorfer** [comment]
Here are two queries that go a long way regarding this topic. Should you decide to add those triples, please remember to remove the `qudt:isScalingOf` triples added in the last PR for the `*GM` units. 

For example, for `unit:KiloGM-PER-DeciM3`, the following triples are generated:
```
   unit:KiloGM-PER-DeciM3
        qudt:isScalingOf  unit:GM-PER-M3 .
        qudt:factorUnit  [ 
             qudt:exponent  1 ;
             qudt:unit unit:KiloGM ] ;
        qudt:factorUnit  [ 
             qudt:exponent  -3 ;
             qudt:unit unit:DeciM ] .
```
A few new properties would need to be defined:
* `qudt:convertsTo` 
* `qudt:factorUnit`
* `qudt:exponent`

The queries are CONSTRUCT queries but they have convenient SELECT  * commented out - use these to get a better overview of what's happening

Note: Both queries will take a little while.

First one:
*  generate `qudt:isScalingOf`.
* If the unit has only one factor, the `qudt:prefix` is generated, too. 
* If the base unit has a conversion multiplier of 1 and a conversion offset of 0 (or missing), also generate `qudt:convertsTo`
```
PREFIX qudt: <http://qudt.org/schema/qudt/>
PREFIX unit: <http://qudt.org/vocab/unit/>
PREFIX prefix: <http://qudt.org/vocab/prefix/>
PREFIX kind: <http://qudt.org/vocab/quantitykind/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
 ?scaledUnit
    qudt:isScalingOf ?baseUnit ;
    qudt:prefix ?prefix ;
    qudt:convertsTo ?convertsTo .
}
#SELECT ?scaledUnit ?baseUnit ?convertsTo ?prefix
where
{
    {
        {

            {
                # select all base units (not using prefixes)
                ?scaledUnit a qudt:Unit ;
                    BIND(REPLACE(STR(?scaledUnit), "^.+/", "") as ?scaledUnitLocalName)
                    FILTER EXISTS {
                        ?prefix
                            a qudt:Prefix ;
                            rdfs:label ?prefixLabel
                        FILTER(CONTAINS(?scaledUnitLocalName, STR(?prefixLabel)))
                    }
            }

            {
                SELECT (CONCAT("\\b(",GROUP_CONCAT(STR(?prefixLabel); separator ="|"), ")") as ?prefixRegex) where
                {
                    ?prefix
                        a qudt:Prefix ;
                        rdfs:label ?prefixLabel .
                 }
            }

        }

        # select all base units (not using prefixes)
        # special handling for prefix + time, e.g.
        # 'unit:MegaYR qudt:isScalingOf unit:YR', not unit:SEC

        ?baseUnit a qudt:Unit ;
        BIND(REPLACE(STR(?baseUnit), "^.+/", "") as ?baseUnitLocalName)
        BIND("(YR|MO|DAY|HR|MIN)\\b" as ?timeRegex)
        FILTER (
                    (
                        REPLACE(?scaledUnitLocalName, CONCAT(?prefixRegex, ?timeRegex), "") = ?scaledUnitLocalName
                        && ?baseUnitLocalName = REPLACE(REPLACE(?scaledUnitLocalName, ?prefixRegex, ""), ?timeRegex, "SEC")
                    ) || (
                        REPLACE(?scaledUnitLocalName, CONCAT(?prefixRegex, ?timeRegex), "") != ?scaledUnitLocalName
                        && ?baseUnitLocalName = REPLACE(?scaledUnitLocalName, ?prefixRegex, "")
                    )
        )
    }
    OPTIONAL {
        ?prefix a qudt:Prefix ;
            rdfs:label ?prefixLabel ;
            FILTER (!CONTAINS(?baseUnitLocalName, "-"))
            FILTER (CONTAINS(?scaledUnitLocalName, STR(?prefixLabel)))
    }
    OPTIONAL {
        ?scaledUnit qudt:conversionMultiplier ?scaledMultiplier . # require any multiplier
        ?baseUnit
            qudt:conversionMultiplier ?cm .
        optional {
            ?baseUnit qudt:conversionOffset ?co
        }
        FILTER( !BOUND(?co) || ?co = 0 )
        FILTER( ?cm = 1 )
        BIND(?baseUnit as ?convertsTo)
    }
}
order by ?baseUnit ?scaledUnit
```

Second: Generate 'factor' unit links, i.e. the combination of base unit and exponent.
Caveat: not all base units exist, e.g. `unit:PicoMOL` only exists in combination with other units, but not on its own. In these cases, the `qudt:factorUnit` triples for that factor unit are missing. One admin task could be to identify missing units, add them, and then run this query again to generate the `qudt:factorUnit` triples.

```
PREFIX qudt: <http://qudt.org/schema/qudt/>
PREFIX unit: <http://qudt.org/vocab/unit/>
PREFIX prefix: <http://qudt.org/vocab/prefix/>
PREFIX kind: <http://qudt.org/vocab/quantitykind/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
    ?derivedUnit qudt:factorUnit [
        qudt:unit ?baseUnit ;
        qudt:exponent ?exponent
    ] .
}
#SELECT * WHERE
#select count(*) where
{
    {
        {
            ?derivedUnit a qudt:Unit ;
                BIND(REPLACE(STR(?derivedUnit), "^.+/", "") as ?derivedUnitLocalName)
            FILTER(REPLACE(?derivedUnitLocalName, "(PER|2|3|4|5|6)\\b", "") != ?derivedUnitLocalName)
        }
        {
            ?baseUnit a qudt:Unit ;
                    BIND(REPLACE(STR(?baseUnit), "^.+/", "") as ?baseUnitLocalName)
                FILTER(!CONTAINS(?baseUnitLocalName, "-"))
                FILTER(REPLACE(?baseUnitLocalName, "(2|3|4|5|6)\\b", "") = ?baseUnitLocalName)
        }
        VALUES (?matchedSuffix) { (1) (2) (3) (4) (5) (6) ("") }
        # replace occurrence with char not found in a localname. we use '@'
        BIND(REPLACE(   ?derivedUnitLocalName,
                        CONCAT("\\b",?baseUnitLocalName, STR(?matchedSuffix),"\\b"),
                        "@")
             AS ?tag)
        # only keep derived units that contain at least one match
        FILTER (?tag != ?derivedUnitLocalName )
        # find out if our match is in the nominator or denominator (or both)
        BIND(REPLACE(REPLACE(?tag, "\\bPER\\b", "/"), "[^@/]", "") as ?fracTagTmp)
        BIND(IF(?fracTagTmp = "@", "@/", ?fracTagTmp) as ?fracTag)
    }
    # match in denominator and nominator separately
    VALUES (?matchFracTag ?expFactor)  { ( "@/" 1 ) ( "/@" -1 ) }
    FILTER (CONTAINS(?fracTag, ?matchFracTag))
    # finally: calculate the exponent
    BIND (?expFactor * IF(?matchedSuffix = "", 1, ?matchedSuffix) as ?exponent)
}
order by ?derivedUnit ?baseUnit
```

All of this information is implicitly contained in the vocabulary, but it's really hard to unearth it. I would appreciate having this information explicitly in the vocabulary.

With these triples added, one might check which `A qudt:isScalingOf B` do not have a `A qudt:convertsTo X` triple because X is not a real base unit but converts to yet another unit, (e.g.  `unit:DecaGM`). Then have a look at `B`, set the `B qudt:convertsTo X` manually, and then transitively set all `A qudt:convertsTo X` triples.

**Comment by fkleedorfer** [comment]
I'll make a PR to sum all of this up. In order for this to work nicely, I need to know how you format your TTL files (otherwise the diff view will be painful).

**Comment by fkleedorfer** [comment]
> Also, I discovered a bug: some units use ..OZ_PER.. instead of ..OZ-PER...
This is now fixed in #458 (if a simple rename is enough)

**Comment by fkleedorfer** [comment]
Closing this as this issue is not focused enough. See #547 and #548.
